### DISCUSSION TITLE
Huge element deformation for phase field fracture

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#21042] [Q&A Modules: General] [answered]
Hello MOOSE Experts

I am trying to solve phase field fracture equation in chemo-mechanical environment. 
Mesh which I am using is structured one.

As soon as the fracture evolves, the local elements deforms enormously (Figure attached).

I am using following solvers

```
solve_type = 'PJFNK'
petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -snes_type'
petsc_options_value = 'hypre boomeramg 300 vinewtonrsls'
automatic_scaling = true
```

Mesh size is resolved as per the length scale of phase field fracture.

![Untitled](https://user-images.githubusercontent.com/94848840/168853477-26d096d9-a813-45dc-9798-26f6b65c51ef.png)

Any thoughts or suggestions?

**Comment by hugary1995** [comment, marked as answer]
This could happen if the residual stiffness of the element is too small. Could you show us the definition of your degradation function?

**Reply by hugary1995** [reply]
The crack band is also much wider than I would expect, which also indicates that there may be some issues in the degradation function, e.g. the derivative of the degradation function w.r.t. damage isn't zero when the material is fully damaged.

**Reply by avtarsinghh1991** [reply]
Hello @hugary1995 

Thanks. Sure, following is the definition of degradation function:

```

[./define_kappa]
type = ParsedMaterial
material_property_names = 'gc_prop l'
f_name = kappa_op
function = 'gc_prop * l'
[../]

[./damage_stress]
type = ComputeLinearElasticPFFractureStress
c = damage
E_name = 'elastic_energy'
D_name = 'degradation'
F_name = 'local_fracture_energy'
decomposition_type = stress_spectral
use_snes_vi_solver = true
[../]

[./degradation]
type = DerivativeParsedMaterial
f_name = degradation
args = 'damage'
function = '(1.0-damage)^2*(1.0 - eta) + eta'
constant_names       = 'eta'
constant_expressions = '1.0e-5'
derivative_order = 2
[../]

[./pfbulkmat]
type = GenericConstantMaterial
prop_names = 'R T cmax ki'
prop_values ='8.314             300           48976               1'
[../]

[./chemical_energy]
type = DerivativeParsedMaterial
f_name = chemical_energy
args = 'diffused'
material_property_names = 'R T cmax ki'
function = 'R*T*cmax*((diffused/cmax)*log(1-diffused/cmax) + (1-diffused/cmax)*log(1-diffused/cmax) + ki*(diffused/cmax)*(1.0-diffused/cmax))'
derivative_order = 2
[../]

[./local_fracture_energy]
type = DerivativeParsedMaterial
f_name = local_fracture_energy
args = 'damage'
material_property_names = 'gc_prop l'
function = 'damage^2 * gc_prop / 2 / l'
derivative_order = 2
[../]

[./fracture_driving_energy]
type = DerivativeSumMaterial
args = 'damage diffused'
sum_materials = 'elastic_energy local_fracture_energy chemical_energy'
derivative_order = 2
f_name = F
[../]
```

Best
Avtar

**Reply by avtarsinghh1991** [reply]
I am also including chemical energy as the driving force of fracture. Do you think it may be the culprit? 
Also, My elastic modulus is the function of concentration.

**Reply by hugary1995** [reply]
Okay, thanks. `eta` is the residual stiffness I was referring to. But usually `eta = 1e-5` is large enough to prevent issues like what you see. It seems that the chemical potential and the fracture potential are decoupled, so I don't see how the formulation could cause such huge deformation. The only other thing I can think of is maybe the solution hasn't really converged? What tolerances are you using?

**Reply by hugary1995** [reply]
> I am also including chemical energy as the driving force of fracture. Do you think it may be the culprit? Also, My elastic modulus is the function of concentration.

But it seems to me your chemical energy is not a function of damage, so I don't think it is actually "driving" fracture. Did I miss something?

**Reply by avtarsinghh1991** [reply]
Yes, chemical energy is not a function of damage. but the diffusivity is the function of damage. And as soon as fracture occurs, the concentration is trapped within the damage zone (Figure attached). As elastic modulus is concentration dependent, I think it may cause the issue.

![Untitled1](https://user-images.githubusercontent.com/94848840/168857648-f75cd755-4a28-4dbb-ba74-38a585ca58f1.png). 


Following is the tolerances which I am using

```
l_max_its = 100
nl_max_its = 30
nl_abs_tol = 5e-3
l_tol = 1e-04
num_steps = 1500
```

**Reply by hugary1995** [reply]
Yeah, just make sure the elastic modulus doesn't drop all the way to zero as concentration increases. The nl_abs_tol is a little bit too high. I usually set `nl_abs_tol = 1e-6` at most, assuming variables are unity scaled.

**Reply by avtarsinghh1991** [reply]
Thanks. Actually, elastic modulus is not becoming zero at ant point of time. I am using following definition of elastic modulus.

```
[./elasticity_tensor_with_Euler_particle]
type = Aniso_Elasticity_Tensor
c = diffused
cmax =  4.8976e4
C1_ijkl='235000000000  127000000000   66000000000  235000000000   66000000000  193000000000   39500000000   39500000000   39500000000'
C0_ijkl='284000000000   70000000000    5800000000  284000000000    5800000000   12500000000    1600000000    1600000000    1600000000'
fill_method1 = symmetric9
fill_method0 = symmetric9
read_prop_user_object = euler_angle_read
block='0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79'
[../]
```

**Reply by avtarsinghh1991** [reply]
Hello @hugary1995 

I put `nl_abs_tol = 1e-6` for my simulation. 

You are right. Its a convergence issue. It's not converging now. 

```
Time Step 251, time = 502, dt = 2
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
 Solve Converged!
  Finished Solving                                                                       [396.00 s] [    0 MB]

Time Step 252, time = 504, dt = 2
  Linear solve did not converge due to DIVERGED_ITS iterations 100
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
 Solve Did NOT Converge!
  Finished Solving                                                                       [133.54 s] [    0 MB]
Aborting as solve did not converge

Time Step 252, time = 503, dt = 1
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100
  Linear solve did not converge due to DIVERGED_ITS iterations 100

```

Any suggestions?

Best
Avtar

**Comment by avtarsinghh1991** [comment]
Hello @hugary1995 

I tried with `nl_abs_tol = 1e-6`
Still the same issue of huge element deformation is coming. 
Length scale for phase field fracture is 4 times the mesh size.
There is anisotropy in the system as well.

I tried with even refined mesh but still no success.

Any other suggestions?

Best
Avtar

**Reply by hugary1995** [reply]
The only other possibility that I can think of is the time step size. I would try reducing the time step size by a factor of 10 and see if the issue persists.

**Reply by avtarsinghh1991** [reply]
Hi @hugary1995 

Thank you very much. I am already running a simulation with reduced time step. 
I will update you with the results soon.

Best
Avtar
